crates/ty_python_semantic/src/types/infer/builder.rs

```diff
@@ -7879,40 +7887,79 @@ impl<'db, 'ast> TypeInferenceBuilder<'db, 'ast> {
             annotation.filter_disjoint_elements(self.db(), collection_ty, inferable)
         });
 
-        // Extract the annotated type of `T`, if provided.
-        let annotated_elt_tys = tcx
-            .known_specialization(self.db(), collection_class)
-            .map(|specialization| specialization.types(self.db()));
+        // Collect type constraints from the declared element types.
+        let (elt_tcx_constraints, elt_tcx_variance) = {
+            let mut builder = SpecializationBuilder::new(
+                self.db(),
+                generic_context.inferable_typevars(self.db()),
+            );
+
+            // For a given type variable, we keep track of the variance of any assignments to
+            // that type variable in the type context.
+            let mut elt_tcx_variance: FxHashMap<BoundTypeVarIdentity<'_>, TypeVarVariance> =
+                FxHashMap::default();
+
+            if let Some(tcx) = tcx.annotation
+                // If there are multiple potential type contexts, we fallback to `Unknown`.
```

Comment:
Where is this fallback actually happening? I don't see it in the code following here.

Reply:
We end up not creating an element constraint, and so will instead infer `Unknown` [a few lines down](https://github.com/astral-sh/ruff/blob/01cd956bc8a150ea92a7d2dc8b9807d6cdd46c97/crates/ty_python_semantic/src/types/infer/builder.rs#L7959).

Reply:
I see -- I think I was also missing that `.class_specialization(self.db()).is_some()` a couple lines below will implicitly exclude unions.